### PR TITLE
change type assignments to interfaces for augmentation purposes

### PR DESCRIPTION
**Pull request by cwadrupldijjit**
## Motivations for this change

I sought to augment the WebSocket implementation for a project I'm working on and I discovered that because of the `type` assignments (such as the `type WebSocket = WebSocketAlias;`) made such a thing impossible.  I imagine some implementations with WS could benefit from type augmentation, similar to how Express handles the `Request` and `Response` objects.

## PR Checklist

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: N/A (implementation of WS is the same, type definition is tweaked slightly with no breaking changes)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header. (Also N/A)

### CODE REVIEW

**Comment by cwadrupldijjit** [comment]
Also another note:  The `dtslint` cli was having issues when I ran it either directly or with `npm test ws`.  It goes up a level from the project root to look for the `notNeededPackages.json`, which naturally doesn't exist there.  Running the test all script, it worked correctly.  I will have to create an issue in the `dtslint` repo to ask about it or suggest a change.

**Review by k-yle — COMMENTED** [review]
Thanks. Do we still need this if #58772 is merged? Also note that #58555 is yet another approach to solving this issue. 

cc @mglyzin and @lavalleeale

**Comment by cwadrupldijjit** [comment]
I'd think both approaches could work simultaneously.  In my case, I don't want to create a custom WebSocket class to use instead.  I just would augment the original WebSocket slightly.

To be fair, I _could_ create a new class, but if it's only for adding a single property, that feels a little much.

But I could also swallow my pride and do it that way instead if that's just how it should be.  Interested to hear others weigh in on this to see if it's necessary.

**Comment by k-yle** [comment]
> But I could also swallow my pride and do it that way instead if that's just how it should be.

The discussion in websockets/ws#2007 suggests that both approaches are valid

> I'd think both approaches could work simultaneously.

I think you're right, we should go ahead with this PR and #58772, which would make #58555 redundant (unless I'm misunderstanding the benefit of that PR?)

**Comment by lavalleeale** [comment]
I think the approach in #58555 would be better unless someone wants to make a single pull request that combines this and #58772?

**Comment by cwadrupldijjit** [comment]
I'd say that the `this` argument in #58555 makes a little more sense from an implementation standpoint, because in theory, that allows the class-based syntax to work fine, too (as seen in #58772).  The only thing that might not be the case is if somehow, the original `WebSocket` implementation is the only value that comes through instead of the custom class.  It also promotes better typings by allowing for the dynamicism of `this` over just the WebSocket implementation.

The result of mine is more related to adding to the original structure of the WebSocket objects instead of creating a new type to handle it.  The `this: this` reference would still handle it just as well as leaving it as `this: WebSocket`, I believe, so it doesn't fully negate the use of it.

I'm also personally for more flexibility than anything.  I don't think any of those approaches are "wrong", though I think that the use cases for each approach might vary.

Prefer to use your own WebSocket class?  That's available for you.  Want to fix the type of the `this` keyword to make more sense?  That's also available for you.  Just want to augment each time a `WebSocket` is used to add minimal properties to it?  That also works.

Does that sound reasonable to you @k-yle and @lavalleeale?

**Comment by cwadrupldijjit** [comment]
Since I anticipate there will be merge conflicts between the three for the tests, I'm willing to merge both of your branches into mine, fix the merge conflicts, and update this PR with those changes, if that makes sense and works for you.  Or I'm willing to be the last one to deal with those in my code if you'd prefer.

**Comment by lavalleeale** [comment]
I think that sounds all good to me as long as the tests reflect all of the use cases

**Comment by vansergen** [comment]
Do we really need tests here?🤔 Otherwise LGTM

File: types/ws/ws-tests.ts
Comment (vansergen) [inline types/ws/ws-tests.ts:330]:
```suggestion
```

Module augmentation is the feature in `TypeScript`. I do not think that there is a need to test explicitly, like

```typescript
declare module 'ws' {
    interface WebSocket {
```

**Review by vansergen — APPROVED** [review]

File: types/ws/ws-tests.ts
Comment (doroodgar) [reply types/ws/ws-tests.ts:330]: ={

Comment (doroodgar) [reply types/ws/ws-tests.ts:330]: =I'd.play

Comment (vansergen) [reply types/ws/ws-tests.ts:330]: > but if it's ever changed back to a type assignment, then that test would fail.

Agree, but the owners of the package would not allow that (I think) because of what we discuss here. I would just leave the test here to explain the motivation of this PR (as an example of usage).

Comment (k-yle) [reply types/ws/ws-tests.ts:330]: I this test is worth keeping in case someone tries to change this back to a `type` assignment. A comment linking to this PR would be helpful but not required.

@cwadrupldijjit, do you want to merge this PR first, as it is now? I'm happy to approve this now and deal with conflicts in the other two PRs

Comment (cwadrupldijjit) [reply types/ws/ws-tests.ts:330]: I'm fine with that.  I don't mind which order things are merged together.  I'm also fine with helping to deal with the merge conflicts.

Comment (vansergen) [reply types/ws/ws-tests.ts:330]: > this test is worth keeping in case someone tries to change this back to a type assignment.

Let's keep it if this is the concern. Is something else blocking this PR from merging?

Comment (vansergen) [reply types/ws/ws-tests.ts:330]: Since we want to keep the test to prevent changing `interface` to `type`, we should add a test for the `Server` class as well (for consistency).

@cwadrupldijjit Could you consider adding it? Something like

```typescript
declare module 'ws' {
    interface Server {
        id?: string;
    }
}
{
    const ws = new wslib.Server({});
    // $ExpectType string | undefined
    ws.id;
    ws.id = 'foo';
}
```

should be enough.

Comment (cwadrupldijjit) [reply types/ws/ws-tests.ts:330]: @vansergen, that's a great point.  Let me add that real quick.

Comment (cwadrupldijjit) [reply types/ws/ws-tests.ts]: I've adjusted the test to include a server augmentation, too.  Could you, @vansergen and @k-yle, review it and see if it looks good to you?

**Review by vansergen — APPROVED** [review]
LGTM

**Review by k-yle — APPROVED** [review]
thanks again

**Review by andrewbranch — APPROVED** [review]

**Comment by cwadrupldijjit** [comment]
Ready to merge

**Pull request merged**
